app/controllers/application_controller.rb

```diff
@@ -1,10 +1,151 @@
-require './config/environment'
-
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "never_trust_the_ide"
+  end
+
+  get '/' do
+    erb :'/index'
+  end
+
+  get '/signup' do
+    if !session[:user_id]
+      erb :'users/create_user', locals: {message: "Please sign up before you sign in"}
+    else
+      redirect to '/tweets'
+    end
+  end
+
+  post '/signup' do
+    if params.any? { |k, v| v.length <=0 }
+      redirect to '/signup'
+    else
+      @user = User.new(:username => params[:username], :email => params[:email], :password => params[:password])
+      @user.save
+      session[:user_id] = @user.id
+      redirect to '/tweets'
+    end
+  end
+
+  get '/login' do
+    if !session[:user_id]
+      erb :'users/login'
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(:username => params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect to '/signup'
+    end
+  end
+
+  get '/logout' do
+    if !!session[:user_id]
+      session.destroy
+      redirect to '/login'
+    else
+      redirect to '/'
+    end
   end
 
-end
\ No newline at end of file
+  get '/tweets' do
+      if session[:user_id]
+        @tweets = Tweet.all
+        erb :'tweets/tweets'
+      else
+        redirect to '/login'
+      end
+    end
+
+    get '/tweets/new' do
+      if session[:user_id]
+        erb :'tweets/create_tweet'
+      else
+        redirect to '/login'
+      end
+    end
+
+    post '/tweets' do
+      if params[:content] == ""
+        redirect to "/tweets/new"
+      else
+        @tweet = Tweet.create(:content => params[:content], :user_id => current_user.id)
+        redirect to "/tweets/#{@tweet.id}"
+      end
+    end
+
+    get '/tweets/:id' do
+      if session[:user_id]
+        @tweet = Tweet.find(params[:id])
+      #  binding.pry
+        erb :'tweets/show_tweet'
+      else
+        redirect to '/login'
+      end
+    end
+
+    get '/tweets/:id/edit' do
+      if session[:user_id]
+        @tweet = Tweet.find(params[:id])
+        if @tweet.user_id == session[:user_id]
+          erb :'tweets/edit_tweet'
+        else
+          redirect to '/tweets'
+        end
+      else
+        redirect to '/login'
+      end
+    end
+
+    post '/tweets/:id' do
+      @tweet = Tweet.find(params[:id])
+      if params[:content] == ""
+        redirect to "/tweets/#{@tweet.id}/edit"
+      else
+        @tweet.content = params[:content]
+        @tweet.save
+        redirect to "/tweets/#{@tweet.id}"
+      end
+    end
+
+    get '/users/:slug' do
+      @user = User.find_by_slug(params[:slug])
+      erb :'users/show'
+    end
+
+    delete '/tweets/:id/delete' do
+      if session[:user_id]
+        @tweet = Tweet.find(params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.